README.md
# Highscores for Risky Cargo game

This project has been done to implement the saving and storing highscores for the game Risky Cargo. The game has been produced as part of IT studies in Tampere University of Applied Sciences.

# Author

Topias Laatu

# Screenshots

![Alt text](/risky_cargo_highscores.png "The website containing the player highscores")

# Tech/framework used

The highscore functions are implemented in Python programming language and the server is running on Flask environment. Gunicorn is used for deploying the app in cloud environment in Render.com. Bcrypt is used for giving a password protection for the http/https requests. The highscores are stored permanently in Firebase.

# Installation and running

!!!
THE CURRENT VERSION OF THE APP DOES NOT WORK AS DESCRIBED BELOW! FIND EARLIER VERSION FROM MAIN BRANCH, WHERE THE FIREBASE SOLUTION IS NOT PRESENT!
!!!

Provide step by step series of examples and explanations about how to get a development env running. Example:

To run this app:

1. Go to a folder where you want to install the app.

2. Open command console and write down the path to the folder.

3. Use this command: 'git clone https://github.com/TobyQuality/highscores-via-python-cloud.git'

4. After the file has been downloaded completely, go to the folder 'highscores-via-python-cloud' (the root directory) with your command console

5. Type the command 'python index.py' (provided you have python installed)

6. Create .env file to your root directory. Inside key value pair in this fashion: API_KEY: "somepassword". Note that API_KEY is without quotation marks and the value, in this case 'somepassword' HAS TO HAVE quotation marks. You can give value you like. It works as the password for the http/https requests.

7. You are good to go! In folder 'test' is a file called 'test.http' where you can specify the requests you make to the server running in your localhost. Remember always to use password attribute like this:  get http://127.0.0.1:5000/api/highscores?pw=secret 

To save your highscores to the cloud environment of the game, you need to wait until the game is finished and downloadable in Google Play store.

# API implementation

API is deployed to cloud and can be accessed using following url:

- https://risky-cargo-highscores.onrender.com/api

# Screencast

[![Screencast](https://upload.wikimedia.org/wikipedia/commons/b/b8/YouTube_Logo_2017.svg)](https://youtu.be/tui9hdY9XlI)

release-notes.txt
RELEASE 1
- Fetching all highscores
- Fetching highscore based on player id
- Fetching highscores with limit (100 is the max limit by default)
- Fetching all scores with ascending or descending order
- Adding a new score
- Deleting highscore (and resetting all level highscores to 0) by player id
- Displaying high scores in html file format in a browser
- Creating new player (account)

RELEASE 2
- Password protection for the requests implemented
- Unit tests for repository module
- Documentation for repository module

RELEASE 3
- Removed player account functionality
- Refactored deleting function
- The highscores webpage now has buttons for limit and sort

RELEASE 4
- Adding highscore from the game has been implemented
- Documentation for the functions
- Tests for functions in index.

RELEASE 5
- Using firebase for permanent storage of highscores

test/test_app.py
import unittest
import requests
import json
from dotenv import load_dotenv
import os

class TestHighscoresAPI(unittest.TestCase):

    load_dotenv()

    API_KEY = os.environ.get('API_KEY')

    API_URL = "http://localhost:5000/api/highscores"

    def test_get_highscores(self):
        # Test getting highscores via API
        response = requests.get(self.API_URL + "?pw=" + self.API_KEY)
        self.assertEqual(response.status_code, 200)

    def test_get_players_highscore(self):
        # Test getting a specific player's highscore via API
        response = requests.get(self.API_URL + "/1?pw=" + self.API_KEY)
        self.assertEqual(response.status_code, 200)

    #def test_add_highscore(self):
        # Test adding a player's highscore via API
        #headers = {'Content-type': 'application/json'}
        #data = json.dumps({'player_name': 'John Doe', 'overall_highscore': 100})
        #response = requests.post(self.API_URL + "?pw=API_KEY", headers=headers, data=data)
        #self.assertEqual(response.status_code, 201)

if __name__ == '__main__':
    unittest.main()

test/test_index.py
from dotenv import load_dotenv    
from index import *
import unittest

class RepositoryTests(unittest.TestCase):

    load_dotenv()
    API_KEY = os.environ.get('API_KEY')

    def test_check_api_key(self):
        """
        Test case for checking API key validation.

        Test case 1: Correct password
        Test case 2: Incorrect password
        """
        # Test case 1: Correct password
        pw = API_KEY
        assert check_api_key(pw) == True

        # Test case 2: Incorrect password
        pw = "wrong_password"
        assert check_api_key(pw) == False
    
    def test_get_highscores(self):
        """
        Test case for getting highscores API endpoint.

        Test case 1: Correct API key
        Test case 2: Incorrect API key
        """
        # Test case 1: Correct API key
        with app.test_client() as client:
            response = client.get('/api/highscores?pw=' + API_KEY)
            assert response.status_code == 200

        # Test case 2: Incorrect API key
        with app.test_client() as client:
            response = client.get('/api/highscores?pw=wrong_password')
            assert response.status_code == 401

    def test_get_players_highscore(self):
        """
        Test case for getting player's highscore API endpoint.

        Test case 1: Correct API key and existing player ID
        Test case 2: Incorrect API key
        Test case 3: Non-existing player ID
        """
        # Test case 1: Correct API key and existing player ID
        with app.test_client() as client:
            response = client.get('/api/highscores/1?pw=' + API_KEY)
            assert response.status_code == 200

        # Test case 2: Incorrect API key
        with app.test_client() as client:
            response = client.get('/api/highscores/1?pw=wrong_password')
            assert response.status_code == 401

        # Test case 3: Non-existing player ID
        with app.test_client() as client:
            response = client.get('/api/highscores/999?pw=' + API_KEY)
            assert response.status_code == 404

    def test_add_highscore(self):
        """
        Test case for adding highscore API endpoint.

        Test case 1: Correct API key and valid data
        Test case 2: Incorrect API key
        Test case 3: Invalid data
        """
        # Test case 1: Correct API key and valid data
        with app.test_client() as client:
            sent_data = {'player_name': 'chummy', 'overall_highscore': 100}
            response = client.post('/api/highscores?pw=' + API_KEY, data=json.dumps(sent_data), content_type='application/json')
            assert response.status_code == 201

        # Test case 2: Incorrect API key
        with app.test_client() as client:
            sent_data = {'player_name': 'dummy', 'overall_highscore': 100}
            response = client.post('/api/highscores?pw=wrong_password', data=json.dumps(sent_data), content_type='application/json')
            assert response.status_code == 401

        # Test case 3: Invalid data
        with app.test_client() as client:
            sent_data = {'player_name': 'd', 'overall_highscore': 100}
            response = client.post('/api/highscores?pw=' + API_KEY, data=json.dumps(sent_data), content_type='application/json')
            assert response.status_code == 400
    
    def test_delete_highscore(self):
        """
        Test case for deleting a highscore API endpoint.

        Test case 1: Correct API key and existing highscore ID
        """
        with app.test_client() as client:
            response = client.delete('/api/highscores/2?pw=' + API_KEY)
            assert response.status_code == 204

test/test_repository.py
import unittest
from utils.repository import *

class RepositoryTests(unittest.TestCase):

    def test_fetch_highscores(self):
        """
        Test case for fetch_highscores function.

        Test fetch_highscores function with limit=5 and sort="desc".
        """
        # Test fetch_highscores function
        highscores = fetch_highscores(limit=5, sort="desc")
        self.assertIsInstance(highscores, list)
        self.assertGreaterEqual(len(highscores), 0)
        self.assertLessEqual(len(highscores), 5)

    def test_fetch_highscore(self):
        """
        Test case for fetch_highscore function.

        Test fetch_highscore function with player ID 1.
        """
        # Test fetch_highscore function
        player_data = fetch_highscore(1)
        self.assertIsInstance(player_data, dict)
        self.assertIn('id', player_data)
        self.assertIn('name', player_data)
        self.assertIn('highscore', player_data)

    def test_save_highscore(self):
        """
        Test case for save_highscore function.

        Test save_highscore function with name="John" and highscore=100.
        """
        # Test save_highscore function
        save_highscore("John", 100)
        player_data = fetch_highscore(2)
        self.assertEqual(player_data['name'], 'John')
        self.assertEqual(player_data['highscore'], 100)

    def test_remove_highscore(self):
        """
        Test case for remove_highscore function.

        Test remove_highscore function with player ID 2.
        """
        #Test remove_highscore function
        remove_highscore(2)
        player_data = fetch_highscore(2)
        self.assertEqual(player_data, None)

    def test_save_highscore_with_invalid_input(self):
        """
        Test case for save_highscore function with invalid input.

        Test save_highscore function with name="John" and highscore="invalid_score".
        Expects an exception to be raised.
        """
        # Test save_highscore function with invalid input
        with self.assertRaises(Exception):
            save_highscore("John", "invalid_score")

utils/repository.py
import json
import os

def open_file(char):
    """
    Open highscores.json file with a given character, which defines the crud-operation (read, write etc.).
    Args:
        char (str): Character representing the crud-operation to be performed on the file ("r" for read, "w" for write, etc.)
    Returns:
        file: File object representing the opened file.
    """
    # in order to reach highscores.json file from folder that is up by one step
    # in the directory, we first have to locate the current folder location of repository.py
    # with os.path.dirname(__file__)
    current_folder = os.path.dirname(__file__)
    # with os.path.join we can find highscores.py with current folder as one argument,
    # ".." as second, which indicates going up the directory by one step,
    # then the file name of highscores as the last argument.
    # Now we have established the relative path to highscores.json
    highscores_file_path = os.path.join(current_folder, '..', 'highscores.json')
    f = open(highscores_file_path, char)
    return f

# This method is used for reading the database, returning
# the content of the file in a string form
def read_database():
    """
    Read the content of the highscores.json file and return it as a string.
    Returns:
        str: Content of the highscores.json file.
    """
    f = open_file("r")
    content = f.read()
    f.close()
    return content.strip()

def fetch_highscores(limit = 100, sort="desc"):
    """
    Fetch the highscores from the highscores.json file.
    Args:
        limit (int, optional): Maximum number of highscores to fetch. Defaults to 100.
        sort (str, optional): Sorting order for highscores ("asc", "desc", or "none"). Defaults to "none".
    Returns:
        list: List of dictionaries containing player names and highscores.
    Raises:
        Exception: If limit is not an integer.
    """
    if type(limit) != int:
        raise Exception("give int")
    # 100 is the default limit
    if limit > 100 or limit <= 0:
        limit = 100
    # if sort is not "none", "asc", or "desc", then it will be "desc"
    # by default, meaning the highscores list will be shown in descending order
    # as opposed to ascending or no particular order
    if sort != "desc":
        if sort != "asc":
            if sort != "none":
                sort = "desc"
    # highscored variable is used for returning a list object.
    # It is specifically needed, because int given as parameter will limit
    # the number of objects that will be returned from json file,
    # so a new list is needed
    highscores = []
    # the loads function converts json to a python object,
    # in this case to a list
    content = json.loads(read_database())
    # this variable determines whether the max range of list objects
    # is the limit or length of the list, dependent on whether the limit
    # is greater than the length of the list or not
    max_range = len(content) if limit > len(content) else limit
    # a list containing dicts will be returned
    if sort == "none":
        return content
    if sort == "asc":
        sorted_highscores = sorted(content, key=lambda d: d['highscore'])
        for i in range(0, max_range):
            player_and_highscore = {}
            player_and_highscore['id'] = content[i]['id']
            player_and_highscore['player'] = sorted_highscores[i]['name']
            player_and_highscore['highscore'] = sorted_highscores[i]['highscore']
            highscores.append(player_and_highscore)
        return highscores
    if sort == "desc":
        sorted_highscores = sorted(content, key=lambda d: d['highscore'], reverse=True)
        for i in range(0, max_range):
            player_and_highscore = {}
            player_and_highscore['id'] = content[i]['id']
            player_and_highscore['player'] = sorted_highscores[i]['name']
            player_and_highscore['highscore'] = sorted_highscores[i]['highscore']
            highscores.append(player_and_highscore)
        return highscores

def fetch_highscore(id):
    """
    Fetch the highscores from the highscores.json file.
    Args:
        limit (int, optional): Maximum number of highscores to fetch. Defaults to 100.
        sort (str, optional): Sorting order for highscores ("asc", "desc", or "none"). Defaults to "none".
    Returns:
        list: List of dictionaries containing player names and highscores.
    Raises:
        Exception: If limit is not an integer.
    """
    if type(id) != int:
        raise Exception("Give integer type of id")
    # if player data remains as none
    # it will be dealt with in index.py
    # with an appropriate response
    player_data = None
    content = json.loads(read_database())
    for data in content:
        if data['id'] == id:
            player_data = data
            break
    return player_data

def save_highscore(name, overall_highscore):
    """
    This function saves the highscore for a specific player in the database.
    
    Args:
        name (str): The name of the player.
        overall_highscore (int): The highscore to be saved.
        
    Returns:
        None
    """
    if type(name) != str:
        raise Exception("Give string object as name")
    if type(overall_highscore) != int:
        raise Exception("Give int as highscore")
    content = json.loads(read_database())
    # new highscore id is initially array length + 1
    id = len(content) + 1
    # check if there is a duplicate id
    duplicate = True
    while duplicate:
        if fetch_highscore(id) != None:
            #print("id was " + str(id))
            id = id + 1
            #print("id is now " + str(id))
        else:
            duplicate = False
    
    new_highscore = {"id": id, "name": name, "highscore": overall_highscore}
    # new highscore is added to the list
    content.append(new_highscore)
    # The list data inside variable 'content' has to be stringified before
    # writing its content to highscores.json file.
    f = open_file("w")
    f.write(json.dumps(content))
    f.close()

def remove_highscore(id):
    """
    This function removes the highscore and for a specific player by ID in the database.
    
    Args:
        id (int): The ID of the player whose highscore is to be deleted.
        
    Returns:
        None
    """
    if type(id) != int:
        raise Exception('Give integer type of id')
    player_data_list = json.loads(read_database())
    # if highscore with given id is found, the highscore will be deleted
    for i in range(len(player_data_list)):
        if player_data_list[i]['id'] == id:
            player_data_list.pop(i)
            break
    f = open_file("w")
    f.write(json.dumps(player_data_list))
    f.close()

def main():
    print("main function is used for testing the module's methods")
    # print(read_database())
    # print(fetch_player_data(1))
    # delete_score_by_id(1)
    print(fetch_highscores(sort="desc"))

if __name__ == "__main__":
    main()

index.py
from flask import *
from json import *
from dotenv import load_dotenv
import os
import bcrypt
import firebase_admin
from firebase_admin import credentials
from firebase_admin import storage
import tempfile

app = Flask(__name__)

# loading the .env file
load_dotenv()

# Credits to Jussi Pohjolainen (pohjus at Github),
# who guided how to set up a permanent storage at Firebase

# read firebase env
# create new variable to render.com named firebase whose
# content is a json file from firebase
json_str = os.environ.get('firebase')

# save the content of the env to a temp file
with tempfile.NamedTemporaryFile(mode='w', delete=False) as f:
    f.write(str(json_str))
    temp_path = f.name

# read json file
cred = credentials.Certificate(temp_path)

# create env to render.com named bucket, whose content is
# e.g.: mydatabase-38cf0.appspot.com
firebase_admin.initialize_app(cred, {
    'storageBucket': os.environ.get('bucket')
})
bucket = storage.bucket()

# accessing the api key used as password with requests to the server
API_KEY = os.environ.get('API_KEY')

# this method is used to check the validity of the password
# sent with requests to the backend
def check_api_key(pw):
    """
    Check the validity of the password sent with requests to the backend.

    Args:
        pw (str): The password string to be checked.

    Returns:
        bool: True if the password is valid, False otherwise.

    Raises:
        None
    """
    # hashing the API_KEY that has been turned to bytes array
    # with randomly generated salt
    hash = bcrypt.hashpw(API_KEY.encode('utf-8'), bcrypt.gensalt())
    # checking password received as parameter with api key
    result = bcrypt.checkpw(pw.encode('utf-8'), hash)
    # returned values are boolean type
    return True if result else False

def fetch_highscores_from_firebase():
    """
    Fetches the highscores data from the Firebase storage bucket.

    Returns:
        dict: The highscores data in dictionary format.
    """
    blob = bucket.blob('highscores.json')
    content = blob.download_as_string().decode('utf-8')
    highscores = json.loads(content)
    return highscores

def fetch_a_highscore_based_on_id(id, highscores):
    """
    Fetches a specific highscore based on the given id from the highscores data.

    Args:
        id (str): The id of the highscore to fetch.
        highscores (dict): The highscores data in dictionary format.

    Returns:
        dict: The highscore data in dictionary format.
    """
    highscore = None
    for data in highscores:
        if data['id'] == id:
            highscore = data
            break
    return highscore

def upload_to_firebase(content, content_type="application/json"):
    """
    Uploads the given content to the Firebase storage bucket.

    Args:
        content (str): The content to upload.
        content_type (str, optional): The content type. Defaults to "application/json".

    Returns:
        None
    """
    blob = bucket.blob('highscores.json')
    blob.upload_from_string(content, content_type)

def validate_name(name):
    """
    Validates a name string.

    Args:
        name (str): The name string to be validated.

    Returns:
        bool: True if the name is valid, False otherwise.

    Raises:
        None

    Example:
        >>> validate_name("John Doe")
        True
        >>> validate_name("A")
        False
        >>> validate_name("This is a very long name")
        False
    """
    if len(name) > 20 or 2 > len(name):
        return False
    return True

@app.route("/")
def root():
    """
    Route for the root URL ("/").

    Returns:
        str: A string with a greeting message.

    Raises:
        None
    """
    return "<h1>:)</h1>"

@app.route("/api/highscores")
def get_highscores():
    """
    Route for getting highscores via API.

    Returns:
        Response: A Flask Response object containing highscores data or an error message.

    Raises:
        None
    """
    pw = request.args.get("pw")
    if (check_api_key(pw) == False):
        return make_response(jsonify("You are unauthorized"), 401)
    highscores = fetch_highscores_from_firebase()
    return make_response(highscores, 200)

@app.route('/api/highscores/<int:the_id>')
def get_players_highscore(the_id):
    """
    Route for getting a specific player's highscore via API.

    Args:
        the_id (int): The ID of the player.

    Returns:
        Response: A Flask Response object containing the player's highscore data or an error message.

    Raises:
        NotFound: If the player with the given ID is not found.
    """
    pw = request.args.get("pw")
    if (check_api_key(pw) == False):
        return make_response(jsonify("You are unauthorized"), 401)
    highscore = fetch_a_highscore_based_on_id(the_id, fetch_highscores_from_firebase())
    if highscore != None:
        return make_response(jsonify(highscore), 200)
    else:
        return make_response(jsonify({'message': 'Player not found.'}), 404)

@app.route("/api/highscores", methods=['POST'])
def add_highscore():
    """
    Route for adding a player's highscore via API.

    Returns:
        Response: A Flask Response object indicating success or failure of the request.

    Raises:
        None
    """
    pw = request.args.get("pw")
    if (check_api_key(pw) == False):
        return make_response(jsonify("You are unauthorized"), 401)
    # the data sent via post request is e.g. {'player_name': 'dummy', 'overall_highscore': 100}
    received_data = json.loads(request.data)
    if validate_name(received_data['player_name']) != True:
        return make_response("Name must be be at least 3 characters and max 20 characters long", 400)
    highscores = fetch_highscores_from_firebase()
    # create id for highscore
    # new highscore id is initially array length + 1
    id = len(highscores) + 1
    # check if there is a duplicate id
    duplicate = True
    while duplicate:
        player_data = fetch_a_highscore_based_on_id(id, highscores)
        if player_data != None:
            id = id + 1
        else:
            duplicate = False
    new_highscore = {"id": id, "name": received_data['player_name'], "highscore": received_data['overall_highscore']}
    # append new highscore to highscores list and jsonify it
    highscores.append(new_highscore)
    highscores_json = json.dumps(highscores)
    # Upload the updated highscores file to Firebase Storage
    upload_to_firebase(content = highscores_json)
    return make_response(jsonify({'message': 'Highscore added successfully!'}), 201)

@app.route('/api/highscores/<int:the_id>', methods=['DELETE'])
def delete_highscore(the_id):
    """
    Deletes a highscore entry for a player with the given ID.

    Args:
        the_id (int): The ID of the player whose highscore is to be deleted.

    Returns:
        Response: A Flask response object with a 204 status code on successful deletion,
        or a response object with 404 status code if the player ID is not found.
    """
    pw = request.args.get("pw")
    if (check_api_key(pw) == False):
        return make_response(jsonify("You are unauthorized"), 401)
    # Before deletion, it has to be made sure that
    # the id of a player exists. Otherwise, a response
    # with status code 404 will be sent
    highscores = fetch_highscores_from_firebase()
    player_data = fetch_a_highscore_based_on_id(the_id, highscores)
    if player_data == None:
        return make_response("", 404)
    for i in range(len(highscores)):
        if highscores[i]['id'] == the_id:
            highscores.pop(i)
            break
    highscores_json = json.dumps(highscores)
    # Upload the updated highscores file to Firebase Storage
    upload_to_firebase(content = highscores_json)
    return make_response("", 204)

@app.route('/highscores')
def show_highscores():
    """
    Renders the highscores page with optional sorting and limiting options.

    Returns:
        Response: A Flask response object containing the rendered highscores page
        with optional sorting and limiting options applied.
    """
    # check, if there is a sort attribute with request
    limit = 0
    sort = ""
    if request.args.get("sort") != None:
        sort = request.args.get("sort")
    # check, if there is a limit attribute with request
    if request.args.get("limit") != None:
        limit = int(request.args.get("limit"))
    #  check, if limit is within 1-100 range, else defaults to 100
    if limit > 100 or limit <= 0:
        limit = 100
    # if sort is not "none", "asc", or "desc", then it will be "desc"
    # by default, meaning the highscores list will be shown in descending order,
    # meaning highest score is first, and so on.
    if sort != "desc":
        if sort != "asc":
            if sort != "none":
                sort = "desc"
    # download highscores from firebase
    highscores = fetch_highscores_from_firebase()
    # this variable determines whether the max range of list objects
    # is the limit or length of the list, dependent on whether the limit
    # is greater than the length of the list or not
    max_range = len(highscores) if limit > len(highscores) else limit
    # a list containing dicts will be returned
    returned_highscores = []
    if sort == "none":
        for i in range(0, max_range):
            returned_highscores.append(highscores[i])
    if sort == "asc":
        sorted_highscores = sorted(highscores, key=lambda d: d['highscore'])
        for i in range(0, max_range):
            returned_highscores.append(sorted_highscores[i])
    if sort == "desc":
        sorted_highscores = sorted(highscores, key=lambda d: d['highscore'], reverse=True)
        for i in range(0, max_range):
            returned_highscores.append(sorted_highscores[i])
    return render_template("highscores.html", highscores=returned_highscores)
    
if __name__ == "__main__":
    app.run(debug=True)